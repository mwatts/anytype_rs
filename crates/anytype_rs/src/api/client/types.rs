//! Type management module
//!
//! Handles type management operations such as creating, updating, and deleting object types.

use super::AnytypeClient;
use crate::{api::types::Icon, error::Result, types::Pagination};
use serde::{Deserialize, Serialize};
use tracing::{debug, info};

/// Property format for type creation
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "snake_case")]
pub enum PropertyFormat {
    Text,
    Number,
    Select,
    MultiSelect,
    Date,
    Files,
    Checkbox,
    Url,
    Email,
    Phone,
    Objects,
}

/// Layout type for object types
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Layout {
    Basic,
    Profile,
    Action,
    Note,
    Bookmark,
    Set,
    Collection,
    Participant,
}

/// Property for type creation
#[derive(Debug, Serialize, Deserialize)]
pub struct CreateTypeProperty {
    pub format: PropertyFormat,
    pub key: String,
    pub name: String,
}

/// Request to create a new type
#[derive(Debug, Serialize)]
pub struct CreateTypeRequest {
    pub icon: Icon,
    pub key: String,
    pub layout: Layout,
    pub name: String,
    pub plural_name: String,
    pub properties: Vec<CreateTypeProperty>,
}

/// Response when creating a type
#[derive(Debug, Deserialize)]
pub struct CreateTypeResponse {
    #[serde(rename = "type")]
    pub type_data: Type,
}

/// Response when getting a type
#[derive(Debug, Deserialize)]
pub struct GetTypeResponse {
    #[serde(rename = "type")]
    pub type_data: Type,
}

/// Property information for types
#[derive(Debug, Deserialize, Serialize)]
pub struct TypeProperty {
    pub format: String,
    pub id: String,
    pub key: String,
    pub name: String,
    pub object: String,
}

/// Type information
#[derive(Debug, Deserialize, Serialize)]
pub struct Type {
    pub archived: Option<bool>,
    pub icon: Icon,
    pub id: String,
    pub key: String,
    pub layout: Option<String>,
    pub name: String,
    pub object: String,
    pub plural_name: Option<String>,
    pub properties: Vec<TypeProperty>,
}

/// Response for listing types
#[derive(Debug, Deserialize)]
pub struct ListTypesResponse {
    pub data: Vec<Type>,
    pub pagination: Pagination,
}

/// Request to update an existing type
#[derive(Debug, Serialize)]
pub struct UpdateTypeRequest {
    pub icon: Icon,
    pub key: String,
    pub layout: Layout,
    pub name: String,
    pub plural_name: String,
    pub properties: Vec<CreateTypeProperty>,
}

/// Response when updating a type
#[derive(Debug, Deserialize)]
pub struct UpdateTypeResponse {
    #[serde(rename = "type")]
    pub type_data: Type,
}

/// Response when deleting a type
#[derive(Debug, Deserialize)]
pub struct DeleteTypeResponse {
    #[serde(rename = "type")]
    pub type_data: Type,
}

impl AnytypeClient {
    /// List types in a space
    pub async fn list_types(&self, space_id: &str) -> Result<Vec<Type>> {
        let response: ListTypesResponse = self.get(&format!("/v1/spaces/{space_id}/types")).await?;
        Ok(response.data)
    }

    /// List types in a space with pagination information
    pub async fn list_types_with_pagination(&self, space_id: &str) -> Result<ListTypesResponse> {
        info!("Listing types in space: {}", space_id);
        debug!("GET /v1/spaces/{}/types", space_id);

        self.get(&format!("/v1/spaces/{space_id}/types")).await
    }

    /// Create a new type in a space
    pub async fn create_type(
        &self,
        space_id: &str,
        request: CreateTypeRequest,
    ) -> Result<CreateTypeResponse> {
        info!("Creating type '{}' in space: {}", request.name, space_id);
        debug!("Request: {:?}", request);
        debug!("Request JSON: {}", serde_json::to_string_pretty(&request)?);

        self.post(&format!("/v1/spaces/{space_id}/types"), &request)
            .await
    }

    /// Get a specific type by ID
    pub async fn get_type(&self, space_id: &str, type_id: &str) -> Result<Type> {
        info!("Getting type '{}' from space: {}", type_id, space_id);
        debug!("GET /v1/spaces/{}/types/{}", space_id, type_id);

        let response: GetTypeResponse = self
            .get(&format!("/v1/spaces/{space_id}/types/{type_id}"))
            .await?;
        Ok(response.type_data)
    }

    /// Update an existing type in a space
    pub async fn update_type(
        &self,
        space_id: &str,
        type_id: &str,
        request: UpdateTypeRequest,
    ) -> Result<UpdateTypeResponse> {
        info!("Updating type '{}' in space: {}", type_id, space_id);
        debug!("Request: {:?}", request);
        debug!("Request JSON: {}", serde_json::to_string_pretty(&request)?);

        self.patch(&format!("/v1/spaces/{space_id}/types/{type_id}"), &request)
            .await
    }

    /// Delete (archive) a type in a space
    pub async fn delete_type(&self, space_id: &str, type_id: &str) -> Result<DeleteTypeResponse> {
        info!(
            "Deleting (archiving) type '{}' in space: {}",
            type_id, space_id
        );
        debug!("DELETE /v1/spaces/{}/types/{}", space_id, type_id);

        self.delete(&format!("/v1/spaces/{space_id}/types/{type_id}"))
            .await
    }
}
