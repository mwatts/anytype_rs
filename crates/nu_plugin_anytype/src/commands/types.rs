use crate::{AnytypePlugin, commands::common::get_space_id, value::AnytypeValue};
use nu_plugin::{EngineInterface, EvaluatedCall, PluginCommand};
use nu_protocol::{Category, LabeledError, PipelineData, Signature, SyntaxShape, Value};

/// Command: anytype type list
pub struct TypeList;

impl PluginCommand for TypeList {
    type Plugin = AnytypePlugin;

    fn name(&self) -> &str {
        "anytype type list"
    }

    fn description(&self) -> &str {
        "List all types in a space"
    }

    fn signature(&self) -> Signature {
        Signature::build(self.name())
            .named(
                "space",
                SyntaxShape::String,
                "Name of the space (can also accept Space from pipeline)",
                Some('s'),
            )
            .input_output_types(vec![
                (
                    nu_protocol::Type::Nothing,
                    nu_protocol::Type::List(Box::new(nu_protocol::Type::Custom(
                        "AnytypeValue".into(),
                    ))),
                ),
                (
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                    nu_protocol::Type::List(Box::new(nu_protocol::Type::Custom(
                        "AnytypeValue".into(),
                    ))),
                ),
            ])
            .category(Category::Custom("anytype".into()))
    }

    fn run(
        &self,
        plugin: &Self::Plugin,
        _engine: &EngineInterface,
        call: &EvaluatedCall,
        input: PipelineData,
    ) -> Result<PipelineData, LabeledError> {
        let span = call.head;
        let input = input.into_value(span)?;

        // Get space_id from multiple sources
        let space_id = get_space_id(plugin, call, &input, span)?;

        // Get client
        let client = plugin.client().map_err(|e| {
            LabeledError::new(format!("Failed to get client: {}", e))
                .with_label("Authentication required", span)
        })?;

        // List types from API
        let types = plugin
            .run_async(client.list_types(&space_id))
            .map_err(|e| LabeledError::new(format!("Failed to list types: {}", e)))?;

        // Convert to AnytypeValue::Type with space_id context
        let values: Vec<Value> = types
            .into_iter()
            .map(|type_data| {
                // Use From<(Type, String)> to convert with space_id context
                let anytype_value: AnytypeValue = (type_data, space_id.clone()).into();
                Value::custom(Box::new(anytype_value), span)
            })
            .collect();

        Ok(PipelineData::Value(Value::list(values, span), None))
    }
}

/// Command: anytype type get
pub struct TypeGet;

impl PluginCommand for TypeGet {
    type Plugin = AnytypePlugin;

    fn name(&self) -> &str {
        "anytype type get"
    }

    fn description(&self) -> &str {
        "Get a type by name"
    }

    fn signature(&self) -> Signature {
        Signature::build(self.name())
            .required("name", SyntaxShape::String, "Name of the type")
            .named(
                "space",
                SyntaxShape::String,
                "Name of the space (can also accept Space from pipeline)",
                Some('s'),
            )
            .input_output_types(vec![
                (
                    nu_protocol::Type::Nothing,
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                ),
                (
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                ),
            ])
            .category(Category::Custom("anytype".into()))
    }

    fn run(
        &self,
        plugin: &Self::Plugin,
        _engine: &EngineInterface,
        call: &EvaluatedCall,
        input: PipelineData,
    ) -> Result<PipelineData, LabeledError> {
        let span = call.head;
        let input = input.into_value(span)?;

        // Get type name from arguments
        let name: String = call.req(0)?;

        // Get space_id from multiple sources
        let space_id = get_space_id(plugin, call, &input, span)?;

        // Get resolver
        let resolver = plugin.resolver().map_err(|e| {
            LabeledError::new(format!("Failed to get resolver: {}", e))
                .with_label("Authentication required", span)
        })?;

        // Resolve type name to ID within the space
        let type_id = plugin
            .run_async(resolver.resolve_type(&space_id, &name))
            .map_err(|e| {
                LabeledError::new(format!(
                    "Failed to resolve type '{}' in space '{}': {}",
                    name, space_id, e
                ))
            })?;

        // Get client
        let client = plugin.client().map_err(|e| {
            LabeledError::new(format!("Failed to get client: {}", e))
                .with_label("Authentication required", span)
        })?;

        // Fetch type details
        let type_data = plugin
            .run_async(client.get_type(&space_id, &type_id))
            .map_err(|e| LabeledError::new(format!("Failed to get type: {}", e)))?;

        // Convert to AnytypeValue::Type with space_id context
        let anytype_value: AnytypeValue = (type_data, space_id).into();
        Ok(PipelineData::Value(
            Value::custom(Box::new(anytype_value), span),
            None,
        ))
    }
}

/// Command: anytype type create
pub struct TypeCreate;

impl PluginCommand for TypeCreate {
    type Plugin = AnytypePlugin;

    fn name(&self) -> &str {
        "anytype type create"
    }

    fn description(&self) -> &str {
        "Create a new type in a space"
    }

    fn signature(&self) -> Signature {
        Signature::build(self.name())
            .required("name", SyntaxShape::String, "Name of the new type")
            .named(
                "key",
                SyntaxShape::String,
                "Unique key for the type",
                Some('k'),
            )
            .named(
                "plural",
                SyntaxShape::String,
                "Plural name for the type",
                Some('p'),
            )
            .named(
                "layout",
                SyntaxShape::String,
                "Layout for the type (basic, profile, action, note, bookmark, set, collection, participant)",
                Some('l'),
            )
            .named(
                "icon",
                SyntaxShape::String,
                "Icon emoji for the type",
                Some('i'),
            )
            .named(
                "space",
                SyntaxShape::String,
                "Name of the space (can also accept Space from pipeline)",
                Some('s'),
            )
            .input_output_types(vec![
                (
                    nu_protocol::Type::Nothing,
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                ),
                (
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                ),
            ])
            .category(Category::Custom("anytype".into()))
    }

    fn run(
        &self,
        plugin: &Self::Plugin,
        _engine: &EngineInterface,
        call: &EvaluatedCall,
        input: PipelineData,
    ) -> Result<PipelineData, LabeledError> {
        let span = call.head;
        let input = input.into_value(span)?;

        // Get arguments
        let name: String = call.req(0)?;
        let key: String = call.get_flag("key")?.ok_or_else(|| {
            LabeledError::new("--key is required").with_label("Missing --key", span)
        })?;
        let plural_name: String = call.get_flag("plural")?.ok_or_else(|| {
            LabeledError::new("--plural is required").with_label("Missing --plural", span)
        })?;
        let layout_str: String = call
            .get_flag("layout")?
            .unwrap_or_else(|| "basic".to_string());
        let icon_emoji: Option<String> = call.get_flag("icon")?;

        // Get space_id from multiple sources
        let space_id = get_space_id(plugin, call, &input, span)?;

        // Parse layout
        let layout = parse_layout(&layout_str).map_err(|e| {
            LabeledError::new(format!("Invalid layout: {}", e))
                .with_label("Invalid layout value", span)
        })?;

        // Parse icon
        let icon = if let Some(emoji) = icon_emoji {
            anytype_rs::Icon::Emoji { emoji }
        } else {
            anytype_rs::Icon::Emoji {
                emoji: "ðŸ“„".to_string(),
            }
        };

        let client = plugin.client().map_err(|e| {
            LabeledError::new(format!("Failed to get client: {}", e))
                .with_label("Authentication required", span)
        })?;

        // Create type with empty properties list (simplified version)
        let request = anytype_rs::CreateTypeRequest {
            key,
            name,
            plural_name,
            layout,
            icon,
            properties: vec![],
        };

        let response = plugin
            .run_async(client.create_type(&space_id, request))
            .map_err(|e| LabeledError::new(format!("Failed to create type: {}", e)))?;

        // Invalidate type cache
        let resolver = plugin
            .resolver()
            .map_err(|e| LabeledError::new(format!("Failed to get resolver: {}", e)))?;
        resolver.clear_cache();

        // Convert to AnytypeValue::Type with space_id context
        let anytype_value: AnytypeValue = (response.type_data, space_id).into();
        Ok(PipelineData::Value(
            Value::custom(Box::new(anytype_value), span),
            None,
        ))
    }
}

/// Command: anytype type update
pub struct TypeUpdate;

impl PluginCommand for TypeUpdate {
    type Plugin = AnytypePlugin;

    fn name(&self) -> &str {
        "anytype type update"
    }

    fn description(&self) -> &str {
        "Update an existing type"
    }

    fn signature(&self) -> Signature {
        Signature::build(self.name())
            .required("name", SyntaxShape::String, "Current name of the type")
            .named(
                "new-name",
                SyntaxShape::String,
                "New name for the type",
                Some('n'),
            )
            .named(
                "key",
                SyntaxShape::String,
                "New key for the type",
                Some('k'),
            )
            .named(
                "plural",
                SyntaxShape::String,
                "New plural name for the type",
                Some('p'),
            )
            .named(
                "layout",
                SyntaxShape::String,
                "New layout for the type",
                Some('l'),
            )
            .named(
                "icon",
                SyntaxShape::String,
                "New icon emoji for the type",
                Some('i'),
            )
            .named(
                "space",
                SyntaxShape::String,
                "Name of the space (can also accept Space from pipeline)",
                Some('s'),
            )
            .input_output_types(vec![
                (
                    nu_protocol::Type::Nothing,
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                ),
                (
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                ),
            ])
            .category(Category::Custom("anytype".into()))
    }

    fn run(
        &self,
        plugin: &Self::Plugin,
        _engine: &EngineInterface,
        call: &EvaluatedCall,
        input: PipelineData,
    ) -> Result<PipelineData, LabeledError> {
        let span = call.head;
        let input = input.into_value(span)?;

        // Get arguments
        let name: String = call.req(0)?;
        let new_name: Option<String> = call.get_flag("new-name")?;
        let key: Option<String> = call.get_flag("key")?;
        let plural_name: Option<String> = call.get_flag("plural")?;
        let layout_str: Option<String> = call.get_flag("layout")?;
        let icon_emoji: Option<String> = call.get_flag("icon")?;

        // Check if at least one update field is provided
        if new_name.is_none()
            && key.is_none()
            && plural_name.is_none()
            && layout_str.is_none()
            && icon_emoji.is_none()
        {
            return Err(
                LabeledError::new("At least one field must be provided to update")
                    .with_label("No update fields provided", span),
            );
        }

        // Get space_id from multiple sources
        let space_id = get_space_id(plugin, call, &input, span)?;

        // Get resolver and client
        let resolver = plugin.resolver().map_err(|e| {
            LabeledError::new(format!("Failed to get resolver: {}", e))
                .with_label("Authentication required", span)
        })?;

        // Resolve type name to ID within the space
        let type_id = plugin
            .run_async(resolver.resolve_type(&space_id, &name))
            .map_err(|e| {
                LabeledError::new(format!(
                    "Failed to resolve type '{}' in space '{}': {}",
                    name, space_id, e
                ))
            })?;

        let client = plugin.client().map_err(|e| {
            LabeledError::new(format!("Failed to get client: {}", e))
                .with_label("Authentication required", span)
        })?;

        // Get current type data to fill in missing fields
        let current_type = plugin
            .run_async(client.get_type(&space_id, &type_id))
            .map_err(|e| LabeledError::new(format!("Failed to get current type: {}", e)))?;

        // Build update request using new values or current values
        let update_name = new_name.unwrap_or(current_type.name);
        let update_key = key.unwrap_or(current_type.key);
        let update_plural = plural_name.unwrap_or(current_type.plural_name.unwrap_or_default());

        let update_layout = if let Some(layout_str) = layout_str {
            parse_layout(&layout_str).map_err(|e| {
                LabeledError::new(format!("Invalid layout: {}", e))
                    .with_label("Invalid layout value", span)
            })?
        } else {
            // Parse current layout or default to basic
            current_type
                .layout
                .as_ref()
                .and_then(|l| parse_layout(l).ok())
                .unwrap_or(anytype_rs::Layout::Basic)
        };

        let update_icon = if let Some(emoji) = icon_emoji {
            anytype_rs::Icon::Emoji { emoji }
        } else {
            current_type.icon
        };

        // Build properties from current type
        let properties: Vec<anytype_rs::CreateTypeProperty> = current_type
            .properties
            .into_iter()
            .map(|p| anytype_rs::CreateTypeProperty {
                format: parse_property_format(&p.format),
                key: p.key,
                name: p.name,
            })
            .collect();

        let request = anytype_rs::UpdateTypeRequest {
            key: update_key,
            name: update_name,
            plural_name: update_plural,
            layout: update_layout,
            icon: update_icon,
            properties,
        };

        let response = plugin
            .run_async(client.update_type(&space_id, &type_id, request))
            .map_err(|e| LabeledError::new(format!("Failed to update type: {}", e)))?;

        // Invalidate type cache
        resolver.invalidate_type(&space_id, &type_id);

        // Convert to AnytypeValue::Type with space_id context
        let anytype_value: AnytypeValue = (response.type_data, space_id).into();
        Ok(PipelineData::Value(
            Value::custom(Box::new(anytype_value), span),
            None,
        ))
    }
}

/// Command: anytype type delete
pub struct TypeDelete;

impl PluginCommand for TypeDelete {
    type Plugin = AnytypePlugin;

    fn name(&self) -> &str {
        "anytype type delete"
    }

    fn description(&self) -> &str {
        "Delete (archive) a type"
    }

    fn signature(&self) -> Signature {
        Signature::build(self.name())
            .required("name", SyntaxShape::String, "Name of the type to delete")
            .named(
                "space",
                SyntaxShape::String,
                "Name of the space (can also accept Space from pipeline)",
                Some('s'),
            )
            .input_output_types(vec![
                (
                    nu_protocol::Type::Nothing,
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                ),
                (
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                ),
            ])
            .category(Category::Custom("anytype".into()))
    }

    fn run(
        &self,
        plugin: &Self::Plugin,
        _engine: &EngineInterface,
        call: &EvaluatedCall,
        input: PipelineData,
    ) -> Result<PipelineData, LabeledError> {
        let span = call.head;
        let input = input.into_value(span)?;

        // Get arguments
        let name: String = call.req(0)?;

        // Get space_id from multiple sources
        let space_id = get_space_id(plugin, call, &input, span)?;

        // Get resolver and client
        let resolver = plugin.resolver().map_err(|e| {
            LabeledError::new(format!("Failed to get resolver: {}", e))
                .with_label("Authentication required", span)
        })?;

        // Resolve type name to ID within the space
        let type_id = plugin
            .run_async(resolver.resolve_type(&space_id, &name))
            .map_err(|e| {
                LabeledError::new(format!(
                    "Failed to resolve type '{}' in space '{}': {}",
                    name, space_id, e
                ))
            })?;

        let client = plugin.client().map_err(|e| {
            LabeledError::new(format!("Failed to get client: {}", e))
                .with_label("Authentication required", span)
        })?;

        // Delete type (archives it)
        let response = plugin
            .run_async(client.delete_type(&space_id, &type_id))
            .map_err(|e| LabeledError::new(format!("Failed to delete type: {}", e)))?;

        // Invalidate type cache
        resolver.invalidate_type(&space_id, &type_id);

        // Convert to AnytypeValue::Type with space_id context
        let anytype_value: AnytypeValue = (response.type_data, space_id).into();
        Ok(PipelineData::Value(
            Value::custom(Box::new(anytype_value), span),
            None,
        ))
    }
}

/// Helper function to parse layout string into Layout enum
fn parse_layout(layout: &str) -> Result<anytype_rs::Layout, String> {
    match layout.to_lowercase().as_str() {
        "basic" => Ok(anytype_rs::Layout::Basic),
        "profile" => Ok(anytype_rs::Layout::Profile),
        "action" => Ok(anytype_rs::Layout::Action),
        "note" => Ok(anytype_rs::Layout::Note),
        "bookmark" => Ok(anytype_rs::Layout::Bookmark),
        "set" => Ok(anytype_rs::Layout::Set),
        "collection" => Ok(anytype_rs::Layout::Collection),
        "participant" => Ok(anytype_rs::Layout::Participant),
        _ => Err(format!(
            "Invalid layout '{}'. Valid options: basic, profile, action, note, bookmark, set, collection, participant",
            layout
        )),
    }
}

/// Helper function to parse property format string into PropertyFormat enum
fn parse_property_format(format: &str) -> anytype_rs::PropertyFormat {
    match format.to_lowercase().as_str() {
        "text" => anytype_rs::PropertyFormat::Text,
        "number" => anytype_rs::PropertyFormat::Number,
        "select" => anytype_rs::PropertyFormat::Select,
        "multi_select" | "multiselect" => anytype_rs::PropertyFormat::MultiSelect,
        "date" => anytype_rs::PropertyFormat::Date,
        "files" => anytype_rs::PropertyFormat::Files,
        "checkbox" => anytype_rs::PropertyFormat::Checkbox,
        "url" => anytype_rs::PropertyFormat::Url,
        "email" => anytype_rs::PropertyFormat::Email,
        "phone" => anytype_rs::PropertyFormat::Phone,
        "objects" => anytype_rs::PropertyFormat::Objects,
        _ => anytype_rs::PropertyFormat::Text, // Default fallback
    }
}
