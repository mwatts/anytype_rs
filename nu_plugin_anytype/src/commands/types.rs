use crate::{commands::common::get_space_id, value::AnytypeValue, AnytypePlugin};
use nu_plugin::{EngineInterface, EvaluatedCall, PluginCommand};
use nu_protocol::{Category, LabeledError, PipelineData, Signature, SyntaxShape, Value};

/// Command: anytype type list
pub struct TypeList;

impl PluginCommand for TypeList {
    type Plugin = AnytypePlugin;

    fn name(&self) -> &str {
        "anytype type list"
    }

    fn description(&self) -> &str {
        "List all types in a space"
    }

    fn signature(&self) -> Signature {
        Signature::build(self.name())
            .named(
                "space",
                SyntaxShape::String,
                "Name of the space (can also accept Space from pipeline)",
                Some('s'),
            )
            .input_output_types(vec![
                (
                    nu_protocol::Type::Nothing,
                    nu_protocol::Type::List(Box::new(nu_protocol::Type::Custom("AnytypeValue".into()))),
                ),
                (
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                    nu_protocol::Type::List(Box::new(nu_protocol::Type::Custom("AnytypeValue".into()))),
                ),
            ])
            .category(Category::Custom("anytype".into()))
    }

    fn run(
        &self,
        plugin: &Self::Plugin,
        _engine: &EngineInterface,
        call: &EvaluatedCall,
        input: PipelineData,
    ) -> Result<PipelineData, LabeledError> {
        let span = call.head;
        let input = input.into_value(span)?;

        // Get space_id from multiple sources
        let space_id = get_space_id(plugin, call, &input, span)?;

        // Get client
        let client = plugin.client().map_err(|e| {
            LabeledError::new(format!("Failed to get client: {}", e))
                .with_label("Authentication required", span)
        })?;

        // List types from API
        let types = plugin
            .run_async(client.list_types(&space_id))
            .map_err(|e| LabeledError::new(format!("Failed to list types: {}", e)))?;

        // Convert to AnytypeValue::Type with space_id context
        let values: Vec<Value> = types
            .into_iter()
            .map(|type_data| {
                // Use From<(Type, String)> to convert with space_id context
                let anytype_value: AnytypeValue = (type_data, space_id.clone()).into();
                Value::custom(Box::new(anytype_value), span)
            })
            .collect();

        Ok(PipelineData::Value(Value::list(values, span), None))
    }
}

/// Command: anytype type get
pub struct TypeGet;

impl PluginCommand for TypeGet {
    type Plugin = AnytypePlugin;

    fn name(&self) -> &str {
        "anytype type get"
    }

    fn description(&self) -> &str {
        "Get a type by name"
    }

    fn signature(&self) -> Signature {
        Signature::build(self.name())
            .required("name", SyntaxShape::String, "Name of the type")
            .named(
                "space",
                SyntaxShape::String,
                "Name of the space (can also accept Space from pipeline)",
                Some('s'),
            )
            .input_output_types(vec![
                (
                    nu_protocol::Type::Nothing,
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                ),
                (
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                    nu_protocol::Type::Custom("AnytypeValue".into()),
                ),
            ])
            .category(Category::Custom("anytype".into()))
    }

    fn run(
        &self,
        plugin: &Self::Plugin,
        _engine: &EngineInterface,
        call: &EvaluatedCall,
        input: PipelineData,
    ) -> Result<PipelineData, LabeledError> {
        let span = call.head;
        let input = input.into_value(span)?;

        // Get type name from arguments
        let name: String = call.req(0)?;

        // Get space_id from multiple sources
        let space_id = get_space_id(plugin, call, &input, span)?;

        // Get resolver
        let resolver = plugin.resolver().map_err(|e| {
            LabeledError::new(format!("Failed to get resolver: {}", e))
                .with_label("Authentication required", span)
        })?;

        // Resolve type name to ID within the space
        let type_id = plugin
            .run_async(resolver.resolve_type(&space_id, &name))
            .map_err(|e| {
                LabeledError::new(format!(
                    "Failed to resolve type '{}' in space '{}': {}",
                    name, space_id, e
                ))
            })?;

        // Get client
        let client = plugin.client().map_err(|e| {
            LabeledError::new(format!("Failed to get client: {}", e))
                .with_label("Authentication required", span)
        })?;

        // Fetch type details
        let type_data = plugin
            .run_async(client.get_type(&space_id, &type_id))
            .map_err(|e| LabeledError::new(format!("Failed to get type: {}", e)))?;

        // Convert to AnytypeValue::Type with space_id context
        let anytype_value: AnytypeValue = (type_data, space_id).into();
        Ok(PipelineData::Value(Value::custom(Box::new(anytype_value), span), None))
    }
}

