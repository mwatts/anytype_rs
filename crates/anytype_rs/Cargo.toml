[package]
name = "anytype_rs"
version.workspace = true
authors.workspace = true
description = "A Rust client library for the Anytype API"
repository.workspace = true
edition.workspace = true
license.workspace = true
keywords = ["anytype", "api", "client"]
categories = ["api-bindings"]
readme = "../../README.md"

[dependencies]
reqwest.workspace = true
serde.workspace = true
serde_json.workspace = true
strum.workspace = true
thiserror.workspace = true
tokio.workspace = true
tracing.workspace = true

[dev-dependencies]
insta.workspace = true
proptest.workspace = true

[lib]
name = "anytype_rs"
path = "src/lib.rs"
