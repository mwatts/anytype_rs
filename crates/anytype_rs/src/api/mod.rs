//! # Anytype API Client
//!
//! A Rust library for interacting with your local Anytype application API.
//!
//! ## Features
//!
//! - Authentication via challenge-response mechanism with local Anytype app
//! - JWT Bearer token support
//! - Full CRUD operations for spaces and objects
//! - Search functionality
//! - Template, type, property, and tag management
//! - Async/await support with tokio
//! - Comprehensive error handling

pub mod client;
pub mod error;
pub mod types;

pub use client::{AnytypeClient, ClientConfig};
pub use error::{AnytypeError, Result};
pub use types::*;

// Re-export types from client modules for convenience
pub use client::auth::{
    CreateApiKeyRequest, CreateApiKeyResponse, CreateChallengeRequest, CreateChallengeResponse,
};
pub use client::lists::{
    AddListObjectsRequest, AddListObjectsResponse, GetListObjectsResponse, GetListViewsResponse,
    ListObject, ListObjectType, ListViewData, ListViewFilter, ListViewSort, ObjectTypeProperty,
    RemoveListObjectsResponse,
};
pub use client::members::{
    GetMemberResponse, ListMembersResponse, Member, MemberRole, MemberStatus,
};
pub use client::objects::{
    CreateObjectRequest, CreateObjectResponse, DeleteObjectResponse, ListObjectsResponse, Object,
    UpdateObjectRequest, UpdateObjectResponse,
};
pub use client::properties::{
    CreatePropertyRequest, CreatePropertyResponse, DeletePropertyResponse, GetPropertyResponse,
    ListPropertiesResponse, Property, UpdatePropertyRequest, UpdatePropertyResponse,
};
pub use client::search::{
    SearchObject, SearchRequest, SearchResponse, SearchSpaceRequest, Sort, SortDirection,
    SortProperty,
};
pub use client::spaces::{
    CreateSpaceRequest, CreateSpaceResponse, ListSpacesResponse, Space, UpdateSpaceRequest,
    UpdateSpaceResponse,
};
pub use client::tags::{
    CreateTagRequest, CreateTagResponse, DeleteTagResponse, GetTagResponse, ListTagsResponse, Tag,
    UpdateTagRequest, UpdateTagResponse,
};
pub use client::templates::{GetTemplateResponse, ListTemplatesResponse, ObjectType, Template};
pub use client::types::{
    CreateTypeProperty, CreateTypeRequest, CreateTypeResponse, DeleteTypeResponse, GetTypeResponse,
    Layout, ListTypesResponse, PropertyFormat, Type, TypeProperty, UpdateTypeRequest,
    UpdateTypeResponse,
};
pub use types::{Icon, IconFormat};
