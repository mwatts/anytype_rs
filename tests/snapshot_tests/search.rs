//! Snapshot tests for search module types

use anytype_rs::api::types::{Icon, Layout, Type};
use anytype_rs::api::{SearchObject, SearchRequest, Sort, SortDirection, SortProperty};

#[test]
fn test_sort_direction_serialization() {
    insta::assert_json_snapshot!("sort_direction_asc", SortDirection::Asc);
    insta::assert_json_snapshot!("sort_direction_desc", SortDirection::Desc);
}

#[test]
fn test_sort_property_serialization() {
    insta::assert_json_snapshot!("sort_property_created_date", SortProperty::CreatedDate);
    insta::assert_json_snapshot!(
        "sort_property_last_modified_date",
        SortProperty::LastModifiedDate
    );
    insta::assert_json_snapshot!(
        "sort_property_last_opened_date",
        SortProperty::LastOpenedDate
    );
    insta::assert_json_snapshot!("sort_property_name", SortProperty::Name);
}

#[test]
fn test_sort_serialization() {
    let sort = Sort {
        direction: SortDirection::Desc,
        property_key: SortProperty::LastModifiedDate,
    };
    insta::assert_json_snapshot!("sort_default", sort);

    let sort_asc_name = Sort {
        direction: SortDirection::Asc,
        property_key: SortProperty::Name,
    };
    insta::assert_json_snapshot!("sort_asc_name", sort_asc_name);
}

#[test]
fn test_search_request_serialization() {
    let request = SearchRequest {
        offset: Some(0),
        limit: Some(50),
        query: Some("test query".to_string()),
        space_id: Some("space123".to_string()),
        sort: Some(Sort {
            direction: SortDirection::Desc,
            property_key: SortProperty::LastModifiedDate,
        }),
    };
    insta::assert_json_snapshot!("search_request_full", request);

    let request_minimal = SearchRequest {
        offset: None,
        limit: None,
        query: None,
        space_id: None,
        sort: None,
    };
    insta::assert_json_snapshot!("search_request_minimal", request_minimal);
}

#[test]
fn test_search_object_serialization() {
    let search_object = SearchObject {
        archived: false,
        icon: Some(Icon::Emoji {
            emoji: "ðŸ“„".to_string(),
        }),
        id: "obj123".to_string(),
        name: "Test Object".to_string(),
        object: "page".to_string(),
        properties: serde_json::json!({
            "title": "Test",
            "description": "A test object"
        }),
        snippet: "This is a snippet...".to_string(),
        space_id: "space456".to_string(),
        r#type: Some(Type {
            archived: false,
            icon: None,
            id: "type789".to_string(),
            key: "page".to_string(),
            layout: Layout::Basic,
            name: "Page".to_string(),
            object: "type".to_string(),
            plural_name: "Pages".to_string(),
            properties: vec![],
        }),
    };
    insta::assert_json_snapshot!("search_object_with_type", search_object);
}
