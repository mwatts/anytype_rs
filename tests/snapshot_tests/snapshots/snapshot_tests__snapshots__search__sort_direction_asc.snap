---
source: tests/snapshot_tests/search.rs
expression: "SortDirection::Asc"
---
"asc"
